internal/dataplane/parser/translate_httproute.go

```diff
@@ -215,31 +213,59 @@ func generateKongRoutesFromHTTPRouteRule(httproute *gatewayv1alpha2.HTTPRoute, r
 
 // generateKongServiceFromHTTPRouteBackendRef converts a provided backendRef for an HTTPRoute
 // into a kong.Service so that routes for that object can be attached to the Service.
-func generateKongServiceFromHTTPRouteBackendRef(result *ingressRules, httproute *gatewayv1alpha2.HTTPRoute, backendRef gatewayv1alpha2.HTTPBackendRef) kongstate.Service {
-	// determine the service namespace
-	// TODO: need to add validation to restrict namespaces in backendRefs
-	namespace := httproute.Namespace
-	if backendRef.Namespace != nil {
-		namespace = string(*backendRef.Namespace)
+func generateKongServiceFromHTTPRouteBackendRef(
+	result *ingressRules,
+	httproute *gatewayv1alpha2.HTTPRoute,
+	ruleNumber int,
+	backendRefs ...gatewayv1alpha2.HTTPBackendRef,
+) (kongstate.Service, error) {
+	// at least one backendRef must be present
+	if len(backendRefs) == 0 {
+		return kongstate.Service{}, fmt.Errorf("no backendRefs present for HTTPRoute %s/%s", httproute.Namespace, httproute.Name)
 	}
 
-	// determine the name of the Service
-	serviceName := fmt.Sprintf("%s.%s.%d", namespace, backendRef.Name, *backendRef.Port)
+	// the first backend found will be used as the anchor for all backends.
+	// this must be done to ensure that the service's host and the upstream
+	// name match, and resolve via DNS due to legacy reasons in the Gateway.
+	anchorBackend := backendRefs[0]
+	anchorNamespace := httproute.Namespace
+	if anchorBackend.Namespace != nil {
+		anchorNamespace = string(*anchorBackend.Namespace)
+	}
+	anchorPort := 80
+	if anchorBackend.Port != nil {
+		anchorPort = int(*anchorBackend.Port)
+	}
 
-	// determine the Service port
-	port := kongstate.PortDef{
-		Mode:   kongstate.PortModeByNumber,
-		Number: int32(*backendRef.Port),
+	// create a kongstate backend for each HTTPRoute backendRef
+	backends := make(kongstate.ServiceBackends, 0, len(backendRefs))
+	for _, backendRef := range backendRefs {
+		// convert each backendRef into a kongstate.ServiceBackend
+		backends = append(backends, kongstate.ServiceBackend{
+			Name: string(backendRef.Name),
+			PortDef: kongstate.PortDef{
+				Mode:   kongstate.PortModeByNumber,
+				Number: int32(*backendRef.Port),
+			},
+			Weight: backendRef.Weight,
+		})
 	}
 
+	// the service name needs to uniquely identify this service given it's list of
+	// one or more backends.
+	serviceName := fmt.Sprintf("%s.%d", getUniqueKongServiceNameForObject(httproute), ruleNumber)
+
+	// the service host needs to be a resolvable name due to legacy logic so we'll
+	// use the anchor backendRef as the basis for the name
+	serviceHost := fmt.Sprintf("%s.%s.%d.svc", anchorBackend.Name, anchorNamespace, anchorPort)
```

Comment:
Ditto here re conflicts. I don't think we get around that without using the rule name (the normal Kong route name minus the match index at the end) and, by extension, changing the name of all upstreams, even those that aren't derived from multi-Service backends. We could use the current Upstream name scheme for backends with only a single Service, but IMO preserving existing upstream names isn't worth the inconsistency in format.

"Resolvable" just means "exists and is a valid hostname"; there's no requirements about specifics beyond that--the `servicename.svc` convention we used originally was arbitrary. `httproute.default.foo.0` or `httproute.default.foo.0.svc` are fine as long as you put that as both the upstream name and service host.

Reply:
Sounds good, 5572595f1b51ad05eb65fb959e719dc27edbe350 changes